File: ruby-sdk/lib/cleanster/api/payment_methods_api.rb
module Cleanster
  module Api
    # PaymentMethodsApi — manage Stripe and PayPal payment methods.
    class PaymentMethodsApi
      def initialize(http)
        @http = http
      end

      # Get Stripe SetupIntent details for collecting card information client-side.
      # @return [Models::ApiResponse]
      def get_setup_intent_details
        raw = @http.get("/v1/payment-methods/setup-intent")
        Models::ApiResponse.from_hash(raw)
      end

      # Get a PayPal client token for rendering the PayPal button client-side.
      # @return [Models::ApiResponse]
      def get_paypal_client_token
        raw = @http.get("/v1/payment-methods/paypal-client-token")
        Models::ApiResponse.from_hash(raw)
      end

      # Save a new payment method after client-side tokenization.
      #
      # @param request [Hash] paymentMethodId (Stripe token) or PayPal nonce.
      # @return [Models::ApiResponse]
      def add_payment_method(request)
        raw = @http.post("/v1/payment-methods", body: request)
        Models::ApiResponse.from_hash(raw)
      end

      # Return all saved payment methods for the current user.
      # @return [Models::ApiResponse]
      def get_payment_methods
        raw = @http.get("/v1/payment-methods")
        Models::ApiResponse.from_hash(raw)
      end

      # Delete a saved payment method.
      #
      # @param payment_method_id [Integer]
      # @return [Models::ApiResponse]
      def delete_payment_method(payment_method_id)
        raw = @http.delete("/v1/payment-methods/#{payment_method_id}")
        Models::ApiResponse.from_hash(raw)
      end

      # Set a payment method as the default for future bookings.
      #
      # @param payment_method_id [Integer]
      # @return [Models::ApiResponse]
      def set_default_payment_method(payment_method_id)
        raw = @http.put("/v1/payment-methods/#{payment_method_id}/default")
        Models::ApiResponse.from_hash(raw)
      end
    end
  end
end

File: php-sdk/src/Api/PaymentMethodsApi.php
<?php

declare(strict_types=1);

namespace Cleanster\Api;

use Cleanster\HttpClient;
use Cleanster\Models\ApiResponse;
use Cleanster\Models\PaymentMethod;

/**
 * Manages Stripe and PayPal payment methods.
 */
final class PaymentMethodsApi
{
    public function __construct(private readonly HttpClient $http) {}

    /**
     * Return Stripe SetupIntent details for client-side card collection.
     * Use the returned clientSecret with Stripe.js on the client side.
     */
    public function getSetupIntentDetails(): ApiResponse
    {
        $raw = $this->http->get('/v1/payment-methods/setup-intent-details');
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Return a PayPal client token for rendering the PayPal button client-side. */
    public function getPaypalClientToken(): ApiResponse
    {
        $raw = $this->http->get('/v1/payment-methods/paypal-client-token');
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /**
     * Save a new payment method after client-side tokenization.
     *
     * @param string $paymentMethodId The tokenized payment method ID from Stripe.js or PayPal.
     */
    public function addPaymentMethod(string $paymentMethodId): ApiResponse
    {
        $raw = $this->http->post('/v1/payment-methods', ['paymentMethodId' => $paymentMethodId]);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /**
     * Return all saved payment methods for the current user.
     *
     * @return ApiResponse<PaymentMethod[]>
     */
    public function getPaymentMethods(): ApiResponse
    {
        $raw   = $this->http->get('/v1/payment-methods');
        $items = array_map(fn(array $m) => new PaymentMethod($m), $raw['data'] ?? []);
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $items);
    }

    /** Remove a saved payment method. */
    public function deletePaymentMethod(int $paymentMethodId): ApiResponse
    {
        $raw = $this->http->delete("/v1/payment-methods/{$paymentMethodId}");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }

    /** Make a saved payment method the default for future bookings. */
    public function setDefaultPaymentMethod(int $paymentMethodId): ApiResponse
    {
        $raw = $this->http->put("/v1/payment-methods/{$paymentMethodId}/default");
        return new ApiResponse($raw['status'] ?? 200, $raw['message'] ?? 'OK', $raw['data'] ?? []);
    }
}
